### PR TITLE
core: DependencyGraph component

### PR DESCRIPTION
**Pull request by zatine**
## Hey, I just made a Pull Request!

We have been using this component internally in multiple places, and this is an attempt to add it as a shared component in open source. Supply nodes (an object with an ID) and edges (objects with to/from, pointing to node IDs) to render a directed graph. It has zooming and panning enabled by default to allow for large and complex graphs.

Some open questions about the implementation:
- This is basically directly copied from our internal implementation, so this relies on a few new dependencies (different [D3](https://github.com/d3) packages and [dagre](https://github.com/dagrejs/dagre) as well as types for these). Are there any preferred packages to use instead, or any functionality that should preferably be omitted to keep the dependencies to a minimum?
- How should the types be exported in the best way? There are some enums like "Direction" that would be beneficial for users to import.
- For ease of use, I went with automatically detecting the size of nodes/labels in order to render the graph properly. Is this a valid assumption or should width/height be supplied by the user with the nodes? (pro would be nicer-looking initial rendering, but would add some overhead for more complicated nodes where width/height might be dynamic) [See examples in this comment.](https://github.com/backstage/backstage/pull/3236#issuecomment-722209354)
- Are there any things that _should_ be customisable but aren't?
- Do the default nodes and edges align well enough with the looks of Backstage? (they need custom implementation due to the graphs rendering in SVG)

Examples from Storybook:

image
image
image

Internal usage example:
image

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation (Storybook)
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)

### CODE REVIEW

**Comment by Fox32** [comment]
Very cool. We already some ideas that could use such a component. Like dependencies between components, APIs, ...

**Comment by Fox32** [comment]
@freben @Rugvip Looking at the checks, is it possible that there are less checks/integrations than before the org migration? I miss the one for storybook, but also codecov?

**Comment by Rugvip** [comment]
@Fox32 Thanks for the heads up, seems like the codecov and chromatic app integration was installed at the Spotify org level and didn't follow along. Installed it in the Backstage org now so expecting them to start showing up.

**Comment by zatine** [comment]
Some more information about dynamically setting width/height vs requiring it for each node.

### Dynamic width/height
Initial rendering looks like this:


Nodes would only need ID for default rendering:
```
const exampleNodes = [
  { id: 'source' },
  { id: 'downstream' },
  { id: 'second-downstream' },
  { id: 'third-downstream' },
];
```

### Static width/height
Initial rendering looks like this:


Nodes would require explicit width/height supplied by the user to render properly:
```
const exampleNodes = [
  { id: 'source', width: 80, height: 50 },
  { id: 'downstream', width: 100, height: 50 },
  { id: 'second-downstream', width: 150, height: 50 },
  { id: 'third-downstream', width: 130, height: 50 },
];
```

File: packages/core/src/components/DependencyGraph/DefaultLabel.tsx [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+function DefaultLabel({ edge: { label } }: RenderLabelProps) {
+  const classes = useStyles();
+  return (
+    <text className={classes.text} textAnchor="middle">
+      {label}
+    </text>
+  );
+}
+export default DefaultLabel;
```
Comment (freben) [inline packages/core/src/components/DependencyGraph/DefaultLabel.tsx:36]: Hi! We have decided to pretty strongly [avoid default exports](https://github.com/backstage/backstage/blob/master/docs/architecture-decisions/adr003-avoid-default-exports.md) - please see rationale in that article. I know it's not a fun ask, but would you mind changing this and all below to be named exports instead?

File: packages/core/src/components/DependencyGraph/DefaultNode.tsx [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      renderedWidth = Math.round(renderedWidth);
+
+      if (renderedHeight !== height || renderedWidth !== width) {
+        setWidth(renderedWidth);
+        setHeight(renderedHeight);
+      }
+    }
+  }, [width, height]);
+
+  const padding = 20;
```
Comment (freben) [inline packages/core/src/components/DependencyGraph/DefaultNode.tsx:55]: don't you want this specified as `10` instead similarly to css? :) and adjust accordingly below

File: packages/core/src/components/DependencyGraph/DependencyGraph.tsx [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+export type DependencyGraphProps = React.SVGProps<SVGSVGElement> & {
+  edges: DependencyEdge[];
+  nodes: DependencyNode[];
+  direction?: Direction;
+  align?: Alignment;
+  nodeMargin?: number;
+  edgeMargin?: number;
+  rankMargin?: number;
+  marginX?: number;
+  marginY?: number;
```
Comment (freben) [inline packages/core/src/components/DependencyGraph/DependencyGraph.tsx:49]: Hm, is this an inner margin that effectively acts as a padding from the outside point of view?

**Review by freben — COMMENTED** [review]
Whoa, this is great stuff! 🎉  Just a few small things

File: packages/core/src/components/DependencyGraph/DefaultLabel.tsx (hunk 1 above)
Comment (zatine) [reply packages/core/src/components/DependencyGraph/DefaultLabel.tsx:36]: Of course!

File: packages/core/src/components/DependencyGraph/DefaultNode.tsx (hunk 2 above)
Comment (zatine) [reply packages/core/src/components/DependencyGraph/DefaultNode.tsx:55]: Yes, that makes sense!

File: packages/core/src/components/DependencyGraph/DependencyGraph.tsx (hunk 3 above)
Comment (zatine) [reply packages/core/src/components/DependencyGraph/DependencyGraph.tsx:49]: Yeah, that's true. I just directly used the options from `dagre`, padding would make more sense to a user. Will rename!

**Comment by freben** [comment]
Sounds great, just ping us again when you have had the time to update things, and I'll re-review!

**Comment by zatine** [comment]
@freben I think I fixed all the comments, please take another look

**Review by freben — APPROVED** [review]
Noooice

**Comment by regicsolutions** [comment]
Not sure if you looked into the React Diagrams project, looks pretty cool - https://github.com/projectstorm/react-diagrams here’s there storybook - https://projectstorm.cloud/react-diagrams/?path=/story/simple-usage--simple-flow-example

**Pull request merged**